README.md
[![Build Status](https://travis-ci.org/bayesjs/bayesjs.svg?branch=master)](https://travis-ci.org/bayesjs/bayesjs)
[![Coverage Status](https://coveralls.io/repos/github/bayesjs/bayesjs/badge.svg)](https://coveralls.io/github/bayesjs/bayesjs)
![npm bundle size](https://img.shields.io/bundlephobia/min/bayesjs)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

# BayesJS

A inference library for [Bayesian Networks](https://en.wikipedia.org/wiki/Bayesian_network) made with [TypeScript](https://www.typescriptlang.org/).

## Inferences

Currently there are three inferences algorithms:

- [Junction tree algorithm](https://en.wikipedia.org/wiki/Junction_tree_algorithm)
- [Variable elimination](https://en.wikipedia.org/wiki/Variable_elimination)
- Enumeration

## Methods

#### infer(network: [INetwork](https://github.com/fhelwanger/bayesjs/blob/master/src/types/INetwork.ts), nodes?: [ICombinations](https://github.com/fhelwanger/bayesjs/blob/master/src/types/ICombinations.ts), given?: [ICombinations](https://github.com/fhelwanger/bayesjs/blob/master/src/types/ICombinations.ts)): number
Calculate the probability of a node's state.

This function receives a network, and will return the probability of the node having the given state subject to the given evidence.

As mentioned above, there are three inferences engines, by default the junction tree algorithm is used to execute the infer function.

If you want to perform repeated inferences from the same bayes network, you should consider instanciating an inference engine for
that network.
```js
import { infer, inferences } from 'bayesjs';

infer(network, event, evidence); // Junction tree algorithm

inferences.enumeration.infer(network, event, evidence);
inferences.variableElimination.infer(network, event, evidence);
inferences.junctionTree.infer(network, event, evidence);
```

##### Example

Given the [rain-sprinkler-grasswet](https://github.com/fhelwanger/bayesjs/blob/master/models/rain-sprinkler-grasswet.ts) network. [Image here](https://en.wikipedia.org/wiki/Bayesian_network#/media/File:SimpleBayesNet.svg).

```js
import { infer } from 'bayesjs';

const network = // ...

// What is the probability that it is raining (RAIN = T)?
infer(network, { 'RAIN': 'T' }).toFixed(4) // 0.2000
// What is the probability that it is raining (RAIN = T), given the sprinkler is off (SPRINKLER = F)?
infer(network, { 'RAIN': 'T' }, { 'SPRINKLER': 'F' }).toFixed(4) // 0.2920
```

#### inferAll(network: [INetwork](https://github.com/fhelwanger/bayesjs/blob/master/src/types/INetwork.ts), given?: [ICombinations](https://github.com/fhelwanger/bayesjs/blob/master/src/types/ICombinations.ts), options?: [IInferAllOptions](https://github.com/fhelwanger/bayesjs/blob/master/src/types/IInferAllOptions.ts)): [INetworkResult](https://github.com/fhelwanger/bayesjs/blob/master/src/types/INetworkResult.ts))
Calculate the marginal distributions for a given network subject to some given evidence.

This method will execute the junction tree algorithm on each node's state.

##### Options

##### precision

default: `8`

Rounds the network results according to this value. To round the value we are using [round-to](https://www.npmjs.com/package/round-to).


Some rounds examples:
- `0.30000000000000004`
  - 8 precision -> `0.3`
  - 4 precision -> `0.3`
  - 2 precision -> `0.3`
- `0.3333333333333333`
  - 8 precision -> `0.33333333`
  - 4 precision -> `0.3333`
  - 2 precision -> `0.33`
- `0.9802979902088171`
  - 8 precision -> `0.98029799`
  - 4 precision -> `0.9803`
  - 2 precision -> `0.98`


##### Example

```js
const network = {
  'Node 1': {
    id: 'Node 1',
    states: ['True', 'False'],
    parents: [],
    cpt: { True: 0.5, False: 0.5 },
  },
  'Node 2': {
    id: 'Node 2',
    states: ['True', 'False'],
    parents: [],
    cpt: { True: 0.5, False: 0.5 },
  },
  'Node 3': {
    id: 'Node 3',
    states: ['True', 'False'],
    parents: ['Node 2', 'Node 1'],
    cpt: [
      {
        when: { 'Node 2': 'True', 'Node 1': 'True' },
        then: { True: 0.5, False: 0.5 },
      },
      {
        when: { 'Node 2': 'False', 'Node 1': 'True' },
        then: { True: 0.5, False: 0.5 },
      },
      {
        when: { 'Node 2': 'True', 'Node 1': 'False' },
        then: { True: 0.5, False: 0.5 },
      },
      {
        when: { 'Node 2': 'False', 'Node 1': 'False' },
        then: { True: 0.5, False: 0.5 },
      },
    ],
  },
};

const evidence = { 'Node 1': 'True' }

inferAll(network, evidence)
// {
//   'Node 1': { True: 1, False: 0 },
//   'Node 2': { True: 0.5, False: 0.5 },
//   'Node 3': { True: 0.5, False: 0.5 },
// }

// Mutating the network...
network["Node 3"].cpt[0].then = { True: 0.95, False: 0.05 };

inferAll(network, evidence);
// {
//   'Node 1': { True: 1, False: 0 },
//   'Node 2': { True: 0.5, False: 0.5 },
//   'Node 3': { True: 0.725, False: 0.275 }
// }
```

#### [IInferenceEngines](https://github.com/fhelwanger/bayesjs/blob/master/src/types/IInferenceEngine.ts))
Inference engines provide an efficient way of performing repeated inferences on the same Bayesian network.  Currently there
is one inference engine provided:

* [InferenceEngine](https://github.com/dlewissandy/bayesjs/blob/feature/inference-engines/src/inferences/junctionTree/hugin-inference-engine.ts)) - An inference engine that uses the junction tree algorithm for infereneces, caching the network topology and potentials for efficiency.

Each engine can be ititialized by calling the constructor with the desired network.  The example below
shows how to perform efficient repeated inference on the Alarm bayes network.:
```js
import { InferenceEngine } from 'bayesjs'
import { allNodes } from './models/alarm.ts'

const network = createNetwork(...allNodes )
const engine = InferenceEngine( network )

// Make multiple inferences with the network without inferences
console.log( engine.infer({ 'JOHN_CALLS': 'T' })) // 0.0521
console.log( engine.infer({ 'MARY_CALLS': 'T' })) // 0.0117
console.log( engine.getEvidence ) // { }

// inject some evidence and make multiple inferences
engine.setEvidence({ 'EARTHQUAKE': 'T' } )
console.log( engine.infer({ 'JOHN_CALLS': 'T' })) // 0.2971
console.log( engine.infer({ 'MARY_CALLS': 'T' })) // 0.2106
console.log( engine.getEvidence ) // { 'EARTHQUAKE': 'T' }

// Update the distribution for a node and make some inferences
engine.setDistribution('BURGLARY', { T: 0.05, F: 0.95 })
console.log( engine.infer({ 'JOHN_CALLS': 'T' })) // 0.3246
console.log( engine.infer({ 'MARY_CALLS': 'T' })) // 0.2329
console.log( engine.getEvidence ) // { 'EARTHQUAKE': 'T' }

// incrementally add additional evidence and make multiple inferences
engine.setEvidence({ 'ALARM': 'T' } )
console.log( engine.infer({ 'JOHN_CALLS': 'T' })) // 0.9000
console.log( engine.infer({ 'MARY_CALLS': 'T' })) // 0.7000
console.log( engine.getEvidence ) // { 'ALARM': 'T', 'EARTHQUAKE': 'T' }

// Remove all the evidence and make multiple inferences.
engine.removeAllEvidence()
console.log( engine.infer({ 'JOHN_CALLS': 'T' })) // 0.0912
console.log( engine.infer({ 'MARY_CALLS': 'T' })) // 0.0435
console.log( engine.getEvidence ) // { }
```

#### addNode(network: [INetwork](https://github.com/fhelwanger/bayesjs/blob/master/src/types/INetwork.ts), node: [INode](https://github.com/fhelwanger/bayesjs/blob/master/src/types/INode.ts)): [INetwork](https://github.com/fhelwanger/bayesjs/blob/master/src/types/INetwork.ts)
Add a node in a Bayesian Network.

This function receives a network and a node, check if the node can be appended on the network. If something is wrong an exception will be thrown, otherwise, a new network will return with the node added.

##### Example

```js
import { addNode } from 'bayesjs';

const networkWithRainAndSprinkler = // ...

const grassWet = {
  id: 'GRASS_WET',
  states: [ 'T', 'F' ],
  parents: [ 'RAIN', 'SPRINKLER' ],
  cpt: [
    { when: { 'RAIN': 'T', 'SPRINKLER': 'T' }, then: { 'T': 0.99, 'F': 0.01 } },
    { when: { 'RAIN': 'T', 'SPRINKLER': 'F' }, then: { 'T': 0.8, 'F': 0.2 } },
    { when: { 'RAIN': 'F', 'SPRINKLER': 'T' }, then: { 'T': 0.9, 'F': 0.1 } },
    { when: { 'RAIN': 'F', 'SPRINKLER': 'F' }, then: { 'T': 0, 'F': 1 } }
  ]
};

const newtwork = addNode(networkWithRainAndSprinkler, grassWet);
```

## License

MIT


CHANGELOG.md
### [0.6.5](https://github.com/bayesjs/bayesjs/compare/v0.6.4...v0.6.5) (2021-09-29)


### :memo: Documentation

* **travis-ci:** update travis ci badge ([53eafb2](https://github.com/bayesjs/bayesjs/commit/53eafb2911c0edfee20fef9067f4c71534b24b3b))


### :bug: Fixes

* **graph:** trivial edit to cause build to restart ([6741e13](https://github.com/bayesjs/bayesjs/commit/6741e1354b020167c35016f20d238b28e9c8f496))
* **infer:** fix inference over junction forests ([f164c8d](https://github.com/bayesjs/bayesjs/commit/f164c8d86f6c835ce77e9cf7a9db4d8824837835))


### :recycle: Refactors

* **potentials:** replace for loop with reduce ([a8cebf3](https://github.com/bayesjs/bayesjs/commit/a8cebf31afda1e0e3c0a776b953b9237cdc26e1e))

### [0.6.4](https://github.com/bayesjs/bayesjs/compare/v0.6.3...v0.6.4) (2021-09-22)


### :bug: Fixes

* **inference:** fix issue with inference returning uniform distribution for some variables ([4c04ca3](https://github.com/bayesjs/bayesjs/commit/4c04ca375d76072953a3505766e1e799a98210f8))

### [0.6.3](https://github.com/bayesjs/bayesjs/compare/v0.6.2...v0.6.3) (2020-04-22)


### :recycle: Refactors

* **graph:** refactor graphs used on junction tree algorithm ([4516b34](https://github.com/bayesjs/bayesjs/commit/4516b349bd39edf1fb945061112f8ff540a5d7c6))

### [0.6.2](https://github.com/bayesjs/bayesjs/compare/v0.6.1...v0.6.2) (2020-04-15)


### :recycle: Refactors

* **junction-tree:** remove unnecessary `flager` helper ([a1afb66](https://github.com/bayesjs/bayesjs/commit/a1afb665bae3d7f4d72b19817bb263fd4e74637c))
* **utils:** remove unnecessary `networkToNodeList` method ([afa15f0](https://github.com/bayesjs/bayesjs/commit/afa15f0e2794cdd13bf9a3939c3de68b249eae8f))

## [0.6.1](https://github.com/bayesjs/bayesjs/compare/v0.6.0...v0.6.1) (2020-04-12)

# [0.6.0](https://github.com/bayesjs/bayesjs/compare/v0.5.0...v0.6.0) (2020-03-30)


### Features

* **infer-all:** add `options` parameter at `inferAll` ([cb9efd1](https://github.com/bayesjs/bayesjs/commit/cb9efd19ea64341d335b2860195346f6069ed931)), closes [#32](https://github.com/bayesjs/bayesjs/issues/32)
* **infer-all:** add `precision` option to round network result values ([faa1b95](https://github.com/bayesjs/bayesjs/commit/faa1b95476917eae6013d4ad8593e1e74b2a48e8))
* **methods:** create net method `inferAll` ([85d4e85](https://github.com/bayesjs/bayesjs/commit/85d4e85087b2e6d07d23c438a03fa088f49b27a8)), closes [#18](https://github.com/bayesjs/bayesjs/issues/18)

# [0.5.0](https://github.com/bayesjs/bayesjs/compare/v0.4.2...v0.5.0) (2020-03-24)


### Features

* **types:** add types in final build ([3376651](https://github.com/bayesjs/bayesjs/commit/33766517b5630b75189d23b71cf99001273e907b))


models/alarm.js
export const burglary = {
  id: 'BURGLARY',
  states: [ 'T', 'F' ],
  parents: [],
  cpt: { 'T': 0.001, 'F': 0.999 }
};

export const earthquake = {
  id: 'EARTHQUAKE',
  states: [ 'T', 'F' ],
  parents: [],
  cpt: { 'T': 0.002, 'F': 0.998 }
};

export const alarm = {
  id: 'ALARM',
  states: [ 'T', 'F' ],
  parents: [ 'BURGLARY', 'EARTHQUAKE' ],
  cpt: [
    { when: { 'BURGLARY': 'T', 'EARTHQUAKE': 'T' }, then: { 'T': 0.95, 'F': 0.05 } },
    { when: { 'BURGLARY': 'T', 'EARTHQUAKE': 'F' }, then: { 'T': 0.94, 'F': 0.06 } },
    { when: { 'BURGLARY': 'F', 'EARTHQUAKE': 'T' }, then: { 'T': 0.29, 'F': 0.71 } },
    { when: { 'BURGLARY': 'F', 'EARTHQUAKE': 'F' }, then: { 'T': 0.001, 'F': 0.999 } }
  ]
};

export const johnCalls = {
  id: 'JOHN_CALLS',
  states: [ 'T', 'F' ],
  parents: [ 'ALARM' ],
  cpt: [
    { when: { 'ALARM': 'T' }, then: { 'T': 0.9, 'F': 0.1 } },
    { when: { 'ALARM': 'F' }, then: { 'T': 0.05, 'F': 0.95 } }
  ]
};

export const maryCalls = {
  id: 'MARY_CALLS',
  states: [ 'T', 'F' ],
  parents: [ 'ALARM' ],
  cpt: [
    { when: { 'ALARM': 'T' }, then: { 'T': 0.7, 'F': 0.3 } },
    { when: { 'ALARM': 'F' }, then: { 'T': 0.01, 'F': 0.99 } }
  ]
};
